Add quotes to column names and disable UpperCase conversion in BaseJdbcClient

agrawalreetika opened the pull request:
## Description
Add quotes to column names and disable UpperCase conversion in BaseJdbcClient

This PR has 2 commits -
1. Disable UpperCase conversion when the case-sensitive flag is enabled
2. Add quotes to column names for renameColumn and dropColumn in BaseJdbcClient

## Motivation and Context
1. **Disable UpperCase conversion when the case-sensitive flag is enabled**
As per the JDBC specification, `storesUpperCaseIdentifiers()` only applies to unquoted identifiers. Since Presto already quotes all identifiers when interacting with JDBC connectors, the result of this method should not influence identifier casing when `caseSensitiveNameMatchingEnabled` is true.

According to the JDBC specification:
```
/**
 * Retrieves whether this database treats mixed case *unquoted* SQL identifiers
 * as case insensitive and stores them in *upper case*.
 */
boolean storesUpperCaseIdentifiers();
```

This PR updates `BaseJdbcClient` to respect the newly introduced `caseSensitiveNameMatchingEnabled` flag when interacting with JDBC metadata.

Presto wraps all identifiers (schemas, tables, columns) with connector-specific delimiters (e.g., "MyTable"), which makes them case-sensitive as per the SQL standard. However, the existing logic blindly uppercases schema and table names when the JDBC DatabaseMetaData.storesUpperCaseIdentifiers() method returns true.

2. Add a quote to column names for renameColumn and dropColumn, as most of the JDBC connectors' case-sensitivity depends on it

## Impact
Improve BaseJdbcClient for case-sensitive scenarios

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

agrawalreetika commented:
> storesUpperCaseIdentifiers

I am very sure, why is it kept in there. 
I kept that as it is to keep the backward compatibility and have the feature as it is when `case-sensitve` flag is not on.

agrawalreetika commented:
My understanding here is that when `caseSensitiveNameMatchingEnabled` we should keep the identifier as it is since Presto wraps the identifier with a delimiter before sending it to the actual data source.

And as `storesUpperCaseIdentifiers` should be considered for the cases when there is no delimiter as per documentation, so `storesUpperCaseIdentifiers` check is breaking `case-sensitve` senarios when `caseSensitiveNameMatchingEnabled` is enabled.
As per my understanding `storesUpperCaseIdentifiers` check should be removed from JDBC as we have delimtier added to identifeir always from Presto for the JDBC-based connector. But I haven't made any changes related to that to keep the Presto legacy behaviour as it is.

BryanCutler commented:
> And as storesUpperCaseIdentifiers should be considered for the cases when there is no delimiter as per documentation, so storesUpperCaseIdentifiers check is breaking case-sensitve senarios when caseSensitiveNameMatchingEnabled is enabled.

But you say Presto will always wrap with a delimiter, so what cases would this be needed?

> As per my understanding storesUpperCaseIdentifiers check should be removed from JDBC as we have delimtier added to identifeir always from Presto for the JDBC-based connector. But I haven't made any changes related to that to keep the Presto legacy behaviour as it is.

It is a change of behavior because if `storesUpperCaseIdentifiers` and `caseSensitiveNameMatchingEnabled` are true, then the identifiers are not capitalized anymore. 

It seems ok to me, if all identifiers are quoted now then the result of `storesUpperCaseIdentifiers` doesn't really matter. 
My only concern is these additions make the checks even more confusing and might be more difficult to straighten out in the future. I'll approve, but I think it would be a good idea to create a github issue about possibly removing `storesUpperCaseIdentifiers` check altogether and make a note of this in the code.

BryanCutler commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleClient.java:
```diff
@@ -115,14 +115,19 @@ protected void renameTable(JdbcIdentity identity, String catalogName, SchemaTabl
+            String sql = format(
+                    "ALTER TABLE %s RENAME TO %s",
+                    quoted(catalogName, oldTable.getSchemaName(), oldTableName),
+                    quoted(newTableName));
```
are the changes here covered in other case-sensitive related tests?

BryanCutler commented on presto-oracle/src/main/java/com/facebook/presto/plugin/oracle/OracleClient.java:
```diff
@@ -115,14 +115,19 @@ protected void renameTable(JdbcIdentity identity, String catalogName, SchemaTabl
+            DatabaseMetaData metadata = connection.getMetaData();
+            String newTableName = newTable.getTableName();
+            String oldTableName = oldTable.getTableName();
+            if (metadata.storesUpperCaseIdentifiers() && !caseSensitiveNameMatchingEnabled) {
```
Was the `storesUpperCaseIdentifiers` just missing for Oracle before or is there something different with it? Please add these Oracle changes explicitly to the PR description.

agrawalreetika replied:
OracleClient had renameTable implementation which had storesUpperCaseIdentifiers missing and converting table names to uppercase by default only for rename operation w/o check. So added it here. I can separate out this in different commit

rschlussel commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -371,7 +371,7 @@ protected JdbcOutputTableHandle createTable(ConnectorTableMetadata tableMetadata
             String remoteSchema = toRemoteSchemaName(session, identity, connection, schemaTableName.getSchemaName());
             String remoteTable = toRemoteTableName(session, identity, connection, remoteSchema, schemaTableName.getTableName());
-            if (uppercase) {
+            if (uppercase && !caseSensitiveNameMatchingEnabled) {
```
why do we have one field for "caseInsensitiveNameMatching" and another "caseSensitiveNameMatchingEnabled". what are they each for?

agrawalreetika replied:
Thanks for the review @rschlussel 

`caseInsensitiveNameMatching` was originally introduced to support case-insensitive matching of schema and table names, useful for databases like MySQL. It works by converting identifiers to lowercase before matching, but this can cause issues when databases contain identifiers that differ only by case (e.g., Test vs TEST).

The new `caseSensitiveNameMatchingEnabled` flag offers broader support for case-sensitive handling of all identifiers — schema, table, and column names — via the `normalizeIdentifier` API. When enabled, Presto preserves the case of user-supplied identifiers, allowing connectors to apply database-specific case sensitivity rules. When disabled, Presto defaults to lowercasing identifiers. More details about this in [here](https://github.com/prestodb/rfcs/pull/36)

agrawalreetika commented:
@BryanCutler PR is now rebased on the latest master also added a description for all 3 commits. Please take a look.
@rschlussel, please review the PR when you get a chance. Thanks!

hantangwangd commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -516,14 +516,14 @@ public void renameColumn(ConnectorSession session, JdbcIdentity identity, JdbcTa
                     quoted(handle.getCatalogName(), handle.getSchemaName(), handle.getTableName()),
-                    jdbcColumn.getColumnName(),
-                    newColumnName);
+                    quoted(jdbcColumn.getColumnName()),
```
Yes, agree with you that adding these conversions into the `storesUpperCaseIdentifiers()` check clause brings no harm, but just make the code more consistent and easier to understand. 

> And about enabling tests for Oracle, its being done as part of #25762

Thanks for the message. Got it!

agrawalreetika replied:
Made changes, please review when you get a chance.
